File: runtimes/py/stream-punk.py
"""Stream-Punk Python Runtime Library
Zero-dependency binary serialization/deserialization for Python.
Compatible with the C++ StreamPunk binary format (little-endian).
"""

import struct
from io import BytesIO
from typing import Optional, Callable, TypeVar, Generic

T = TypeVar('T')


class SpRef(Generic[T]):
    def __init__(self, value: Optional[T], address: int):
        self.value = value
        self.address = address


class SpArray(Generic[T]):
    def __init__(self, size: int, initializer: T = None):
        self._data = [initializer] * size

    def size(self) -> int:
        return len(self._data)

    def at(self, index: int) -> T:
        if index < 0 or index >= len(self._data):
            raise IndexError("Index out of bounds")
        return self._data[index]

    def set(self, index: int, value: T):
        if index < 0 or index >= len(self._data):
            raise IndexError("Index out of bounds")
        self._data[index] = value


class SpVariant:
    def __init__(self, value=None):
        self._value = value
        self._type_index = -1
        if value is not None:
            self._update_type_index()

    def set(self, value):
        self._value = value
        self._update_type_index()

    @property
    def value(self):
        return self._value

    @property
    def type_index(self) -> int:
        return self._type_index

    def _update_type_index(self):
        self._type_index = -1

    def possible_types(self):
        return []


class I:
    def __init__(self, data: bytes, offset: int = 0):
        self._buf = data
        self.off = offset
        self._obj_map: dict[int, object] = {}

    def has_more_data(self) -> bool:
        return self.off < len(self._buf)

    def read_u8(self) -> int:
        v = self._buf[self.off]
        self.off += 1
        return v

    def read_u16(self) -> int:
        v = struct.unpack_from('<H', self._buf, self.off)[0]
        self.off += 2
        return v

    def read_u32(self) -> int:
        v = struct.unpack_from('<I', self._buf, self.off)[0]
        self.off += 4
        return v

    def read_u64(self) -> int:
        v = struct.unpack_from('<Q', self._buf, self.off)[0]
        self.off += 8
        return v

    def read_i8(self) -> int:
        v = struct.unpack_from('<b', self._buf, self.off)[0]
        self.off += 1
        return v

    def read_i16(self) -> int:
        v = struct.unpack_from('<h', self._buf, self.off)[0]
        self.off += 2
        return v

    def read_i32(self) -> int:
        v = struct.unpack_from('<i', self._buf, self.off)[0]
        self.off += 4
        return v

    def read_i64(self) -> int:
        v = struct.unpack_from('<q', self._buf, self.off)[0]
        self.off += 8
        return v

    def read_f32(self) -> float:
        v = struct.unpack_from('<f', self._buf, self.off)[0]
        self.off += 4
        return v

    def read_f64(self) -> float:
        v = struct.unpack_from('<d', self._buf, self.off)[0]
        self.off += 8
        return v

    def read_ch(self) -> str:
        v = self._buf[self.off]
        self.off += 1
        return chr(v)

    def read_ch8(self) -> str:
        v = self._buf[self.off]
        self.off += 1
        return chr(v)

    def read_ch16(self) -> str:
        v = struct.unpack_from('<H', self._buf, self.off)[0]
        self.off += 2
        return chr(v)

    def read_ch32(self) -> str:
        cp = struct.unpack_from('<I', self._buf, self.off)[0]
        self.off += 4
        if 0 <= cp <= 0x10FFFF:
            return chr(cp)
        return ''

    def read_bl(self) -> bool:
        v = self._buf[self.off] != 0
        self.off += 1
        return v

    def read_sz(self) -> int:
        return self.read_u32()

    def read_bytes(self, length: int):
        self.off += length

    def read_string(self) -> str:
        length = self.read_sz()
        if length == 0:
            return ''
        v = self._buf[self.off:self.off + length].decode('utf-8')
        self.off += length
        return v

    def read_u8string(self) -> str:
        length = self.read_sz()
        v = self._buf[self.off:self.off + length].decode('utf-8')
        self.off += length
        return v

    def read_u16string(self) -> str:
        length = self.read_sz()
        byte_len = length * 2
        v = self._buf[self.off:self.off + byte_len].decode('utf-16-le')
        self.off += byte_len
        return v

    def read_u32string(self) -> bytes:
        length = self.read_sz()
        byte_len = length * 4
        v = self._buf[self.off:self.off + byte_len]
        self.off += byte_len
        return v

    def read_ptr_with_typeID(self):
        addr = self.read_u64()
        if addr == 0:
            return SpRef(None, 0)
        if addr in self._obj_map:
            return self._obj_map[addr]
        ref = SpRef(None, addr)
        self._obj_map[addr] = ref
        ref.value = read_obj(self)
        return ref

    def read_ptr(self, reader: Callable[[], T]) -> SpRef[T]:
        addr = self.read_u64()
        if addr == 0:
            return SpRef(None, 0)
        if addr in self._obj_map:
            return self._obj_map[addr]
        value = reader()
        ref = SpRef(value, addr)
        self._obj_map[addr] = ref
        return ref

    def read_array(self, count: int, reader: Callable[[], T]) -> list:
        return [reader() for _ in range(count)]

    def read_Array(self, reader: Callable[[], T]) -> list:
        size = self.read_sz()
        return self.read_array(size, reader)

    def read_set(self, reader: Callable[[], T]) -> set:
        arr = self.read_Array(reader)
        try:
            return set(arr)
        except TypeError:
            return set(tuple(x) for x in arr)

    def read_unordered_set(self, reader: Callable[[], T]) -> set:
        return self.read_set(reader)

    def read_map(self, key_reader: Callable, value_reader: Callable) -> dict:
        size = self.read_sz()
        result = {}
        for _ in range(size):
            k = key_reader()
            v = value_reader()
            result[k] = v
        return result

    def read_unordered_map(self, key_reader: Callable, value_reader: Callable) -> dict:
        return self.read_map(key_reader, value_reader)

    def read_vector(self, reader: Callable[[], T]) -> list:
        return self.read_Array(reader)

    def read_deque(self, reader: Callable[[], T]) -> list:
        return self.read_Array(reader)

    def read_list(self, reader: Callable[[], T]) -> list:
        return self.read_Array(reader)

    def read_forward_list(self, reader: Callable[[], T]) -> list:
        return self.read_Array(reader)

    def read_SpArray(self, size: int, reader: Callable[[], T]) -> SpArray[T]:
        arr = SpArray(size)
        for i in range(size):
            arr.set(i, reader())
        return arr

    def read_std_string(self) -> str:
        return self.read_string()

    def read_bitset(self) -> list:
        size = self.read_u32()
        byte_len = (size + 7) // 8
        result = []
        for i in range(size):
            byte_idx = i // 8
            bit_idx = i % 8
            byte_val = self._buf[self.off + byte_idx]
            result.append(((byte_val >> bit_idx) & 1) == 1)
        self.off += byte_len
        return result

    def read_optional(self, reader: Callable[[], T]) -> Optional[T]:
        if self.read_bl():
            return reader()
        return None

    def read_variant(self, readers: list) -> object:
        index = self.read_u32()
        if index < 0 or index >= len(readers):
            raise IndexError(f"Variant index {index} out of range")
        return readers[index]()

    def read_stream_punk_time(self) -> int:
        sec = self.read_i64()
        atto_sec = self.read_i64()
        return sec * 1000 + atto_sec // 10**15


class O:
    def __init__(self):
        self._buf = BytesIO()
        self._obj_map: dict[int, object] = {}

    def write_u8(self, v: int):
        self._buf.write(struct.pack('<B', v))

    def write_u16(self, v: int):
        self._buf.write(struct.pack('<H', v))

    def write_u32(self, v: int):
        self._buf.write(struct.pack('<I', v))

    def write_u64(self, v: int):
        self._buf.write(struct.pack('<Q', v))

    def write_i8(self, v: int):
        self._buf.write(struct.pack('<b', v))

    def write_i16(self, v: int):
        self._buf.write(struct.pack('<h', v))

    def write_i32(self, v: int):
        self._buf.write(struct.pack('<i', v))

    def write_i64(self, v: int):
        self._buf.write(struct.pack('<q', v))

    def write_f32(self, v: float):
        self._buf.write(struct.pack('<f', v))

    def write_f64(self, v: float):
        self._buf.write(struct.pack('<d', v))

    def write_ch(self, v: str):
        self.write_ch16(v)

    def write_ch8(self, v: str):
        if len(v) > 0:
            self.write_u8(ord(v[0]))
        else:
            self.write_u8(0)

    def write_ch16(self, v: str):
        if len(v) > 0:
            self.write_u16(ord(v[0]))
        else:
            self.write_u16(0)

    def write_ch32(self, v: str):
        if len(v) > 0:
            self.write_u32(ord(v[0]))
        else:
            self.write_u32(0)

    def write_bl(self, v: bool):
        self.write_u8(1 if v else 0)

    def write_sz(self, v: int):
        self.write_u32(v)

    def write_string(self, v: str):
        encoded = v.encode('utf-8')
        self.write_sz(len(encoded))
        self._buf.write(encoded)

    def write_u8string(self, v: str):
        encoded = v.encode('utf-8')
        self.write_sz(len(encoded))
        self._buf.write(encoded)

    def write_u16string(self, v: str):
        encoded = v.encode('utf-16-le')
        self.write_sz(len(v))
        self._buf.write(encoded)

    def write_u32string(self, v: bytes):
        self.write_sz(len(v) // 4)
        self._buf.write(v)

    def write_ptr(self, value, address: int, writer: Callable):
        if value is None:
            self.write_u64(0)
            return
        if address == 0:
            address = len(self._obj_map) + 1
        if address in self._obj_map:
            self.write_u64(address)
            return
        self._obj_map[address] = value
        self.write_u64(address)
        writer(value)

    def write_ptr_with_typeID(self, value, address: int = 0):
        if value is None:
            self.write_u64(0)
            return
        if address == 0:
            address = len(self._obj_map) + 1
        if address in self._obj_map:
            self.write_u64(address)
            return
        self._obj_map[address] = value
        self.write_u64(address)
        write_obj(self, value)

    def write_array(self, arr: list, writer: Callable):
        for item in arr:
            writer(item)

    def write_Array(self, arr: list, writer: Callable):
        self.write_sz(len(arr))
        self.write_array(arr, writer)

    def write_set(self, s: set, writer: Callable):
        self.write_Array(list(s), writer)

    def write_unordered_set(self, s: set, writer: Callable):
        self.write_set(s, writer)

    def write_map(self, d: dict, key_writer: Callable, value_writer: Callable):
        self.write_sz(len(d))
        for k, v in d.items():
            key_writer(k)
            value_writer(v)

    def write_unordered_map(self, d: dict, key_writer: Callable, value_writer: Callable):
        self.write_map(d, key_writer, value_writer)

    def write_vector(self, vec: list, writer: Callable):
        self.write_Array(vec, writer)

    def write_deque(self, deq: list, writer: Callable):
        self.write_Array(deq, writer)

    def write_list(self, lst: list, writer: Callable):
        self.write_Array(lst, writer)

    def write_forward_list(self, lst: list, writer: Callable):
        self.write_Array(lst, writer)

    def write_SpArray(self, arr: SpArray, writer: Callable):
        for i in range(arr.size()):
            writer(arr.at(i))

    def write_bitset(self, bits: list):
        self.write_u32(len(bits))
        byte_len = (len(bits) + 7) // 8
        bytes_arr = bytearray(byte_len)
        for i, bit in enumerate(bits):
            if bit:
                byte_idx = i // 8
                bit_idx = i % 8
                bytes_arr[byte_idx] |= (1 << bit_idx)
        self._buf.write(bytes_arr)

    def write_optional(self, value, writer: Callable):
        self.write_bl(value is not None)
        if value is not None:
            writer(value)

    def write_variant(self, value, index: int, writers: list):
        self.write_u32(index)
        writers[index](value)

    def write_stream_punk_time(self, t_ms: int):
        sec = t_ms // 1000
        atto_sec = (t_ms % 1000) * 10**15
        self.write_i64(sec)
        self.write_i64(atto_sec)

    def to_bytes(self) -> bytes:
        return self._buf.getvalue()

File: runtimes/go/stream-punk.go
/**
 * Stream-Punk Go Runtime Library
 * Zero-dependency binary serialization/deserialization for Go.
 * Compatible with the C++ StreamPunk binary format (little-endian).
 */

package main

import (
	"encoding/binary"
	"math"
	"unicode/utf16"
)

/* =================== 辅助类型 =================== */

type SpRef struct {
	Value   interface{}
	Address uint64
}

func NewSpRef() SpRef {
	return SpRef{Value: nil, Address: 0}
}

type SpArray struct {
	Data []interface{}
}

func NewSpArray(size int) SpArray {
	return SpArray{Data: make([]interface{}, size)}
}

func (a *SpArray) At(index int) interface{} {
	return a.Data[index]
}

func (a *SpArray) Set(index int, value interface{}) {
	a.Data[index] = value
}

func (a *SpArray) Size() int {
	return len(a.Data)
}

type SpVariant struct {
	TypeIndex uint32
	Value     interface{}
}

func NewSpVariant() SpVariant {
	return SpVariant{TypeIndex: 0, Value: nil}
}

func (v *SpVariant) WriteVariant(o *O) {
	o.WriteU32(v.TypeIndex)
	if v.Value != nil {
		if obj, ok := v.Value.(SpBase); ok {
			WriteObj(o, obj)
		}
	}
}

func ReadVariant(i *I) SpVariant {
	typeIndex := i.ReadU32()
	value := ReadObj(i)
	return SpVariant{TypeIndex: typeIndex, Value: value}
}

/* =================== SpBase 接口 =================== */

type SpBase interface {
	TypeID() uint32
	From_(i *I)
	To(o *O)
}

/* =================== 反序列化读取器 =================== */

type I struct {
	buf    []byte
	Off    int
	objMap map[uint64]interface{}
}

func NewI(data []byte) *I {
	return &I{buf: data, Off: 0, objMap: make(map[uint64]interface{})}
}

func NewIOffset(data []byte, offset int) *I {
	return &I{buf: data, Off: offset, objMap: make(map[uint64]interface{})}
}

func (i *I) HasMoreData() bool {
	return i.Off < len(i.buf)
}

func (i *I) ReadU8() uint8 {
	v := i.buf[i.Off]
	i.Off++
	return v
}

func (i *I) ReadU16() uint16 {
	v := binary.LittleEndian.Uint16(i.buf[i.Off:])
	i.Off += 2
	return v
}

func (i *I) ReadU32() uint32 {
	v := binary.LittleEndian.Uint32(i.buf[i.Off:])
	i.Off += 4
	return v
}

func (i *I) ReadU64() uint64 {
	v := binary.LittleEndian.Uint64(i.buf[i.Off:])
	i.Off += 8
	return v
}

func (i *I) ReadI8() int8 {
	v := int8(i.buf[i.Off])
	i.Off++
	return v
}

func (i *I) ReadI16() int16 {
	v := int16(binary.LittleEndian.Uint16(i.buf[i.Off:]))
	i.Off += 2
	return v
}

func (i *I) ReadI32() int32 {
	v := int32(binary.LittleEndian.Uint32(i.buf[i.Off:]))
	i.Off += 4
	return v
}

func (i *I) ReadI64() int64 {
	v := int64(binary.LittleEndian.Uint64(i.buf[i.Off:]))
	i.Off += 8
	return v
}

func (i *I) ReadF32() float32 {
	bits := binary.LittleEndian.Uint32(i.buf[i.Off:])
	i.Off += 4
	return math.Float32frombits(bits)
}

func (i *I) ReadF64() float64 {
	bits := binary.LittleEndian.Uint64(i.buf[i.Off:])
	i.Off += 8
	return math.Float64frombits(bits)
}

func (i *I) ReadCh() byte {
	v := i.buf[i.Off]
	i.Off++
	return v
}

func (i *I) ReadCh8() byte {
	v := i.buf[i.Off]
	i.Off++
	return v
}

func (i *I) ReadCh16() uint16 {
	v := binary.LittleEndian.Uint16(i.buf[i.Off:])
	i.Off += 2
	return v
}

func (i *I) ReadCh32() uint32 {
	v := binary.LittleEndian.Uint32(i.buf[i.Off:])
	i.Off += 4
	return v
}

func (i *I) ReadBl() bool {
	v := i.buf[i.Off]
	i.Off++
	return v != 0
}

func (i *I) ReadSz() int {
	return int(i.ReadU32())
}

func (i *I) ReadBytes(length int) {
	i.Off += length
}

func (i *I) ReadString() string {
	length := i.ReadSz()
	if length == 0 {
		return ""
	}
	v := string(i.buf[i.Off : i.Off+length])
	i.Off += length
	return v
}

func (i *I) ReadU8String() string {
	return i.ReadString()
}

func (i *I) ReadU16String() string {
	length := i.ReadSz()
	if length == 0 {
		return ""
	}
	byteLen := length * 2
	u16s := make([]uint16, length)
	for j := 0; j < length; j++ {
		u16s[j] = binary.LittleEndian.Uint16(i.buf[i.Off+j*2:])
	}
	i.Off += byteLen
	return string(utf16.Decode(u16s))
}

func (i *I) ReadU32String() []uint32 {
	length := i.ReadSz()
	byteLen := length * 4
	result := make([]uint32, length)
	for j := 0; j < length; j++ {
		result[j] = binary.LittleEndian.Uint32(i.buf[i.Off+j*4:])
	}
	i.Off += byteLen
	return result
}

func (i *I) ReadPtrWithTypeID() SpRef {
	addr := i.ReadU64()
	if addr == 0 {
		return NewSpRef()
	}
	if ref, ok := i.objMap[addr]; ok {
		return ref.(SpRef)
	}
	ref := SpRef{Value: nil, Address: addr}
	i.objMap[addr] = ref
	ref.Value = ReadObj(i)
	i.objMap[addr] = ref
	return ref
}

func (i *I) ReadPtr(reader func() interface{}) SpRef {
	addr := i.ReadU64()
	if addr == 0 {
		return NewSpRef()
	}
	if ref, ok := i.objMap[addr]; ok {
		return ref.(SpRef)
	}
	ref := SpRef{Value: nil, Address: addr}
	i.objMap[addr] = ref
	ref.Value = reader()
	i.objMap[addr] = ref
	return ref
}

func (i *I) ReadArray(reader func() interface{}) []interface{} {
	length := i.ReadSz()
	result := make([]interface{}, length)
	for j := 0; j < length; j++ {
		result[j] = reader()
	}
	return result
}

func (i *I) ReadSet(reader func() interface{}) map[interface{}]struct{} {
	length := i.ReadSz()
	result := make(map[interface{}]struct{})
	for j := 0; j < length; j++ {
		result[reader()] = struct{}{}
	}
	return result
}

func (i *I) ReadMap(keyReader func() interface{}, valueReader func() interface{}) map[interface{}]interface{} {
	length := i.ReadSz()
	result := make(map[interface{}]interface{})
	for j := 0; j < length; j++ {
		result[keyReader()] = valueReader()
	}
	return result
}

func (i *I) ReadSpArray(size int, reader func() interface{}) SpArray {
	arr := NewSpArray(size)
	for j := 0; j < size; j++ {
		arr.Set(j, reader())
	}
	return arr
}

/* =================== 序列化写入器 =================== */

type O struct {
	buf []byte
}

func NewO() *O {
	return &O{buf: make([]byte, 0, 256)}
}

func (o *O) Bytes() []byte {
	return o.buf
}

func (o *O) WriteU8(v uint8) {
	o.buf = append(o.buf, v)
}

func (o *O) WriteU16(v uint16) {
	b := make([]byte, 2)
	binary.LittleEndian.PutUint16(b, v)
	o.buf = append(o.buf, b...)
}

func (o *O) WriteU32(v uint32) {
	b := make([]byte, 4)
	binary.LittleEndian.PutUint32(b, v)
	o.buf = append(o.buf, b...)
}

func (o *O) WriteU64(v uint64) {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, v)
	o.buf = append(o.buf, b...)
}

func (o *O) WriteI8(v int8) {
	o.buf = append(o.buf, byte(v))
}

func (o *O) WriteI16(v int16) {
	b := make([]byte, 2)
	binary.LittleEndian.PutUint16(b, uint16(v))
	o.buf = append(o.buf, b...)
}

func (o *O) WriteI32(v int32) {
	b := make([]byte, 4)
	binary.LittleEndian.PutUint32(b, uint32(v))
	o.buf = append(o.buf, b...)
}

func (o *O) WriteI64(v int64) {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, uint64(v))
	o.buf = append(o.buf, b...)
}

func (o *O) WriteF32(v float32) {
	b := make([]byte, 4)
	binary.LittleEndian.PutUint32(b, math.Float32bits(v))
	o.buf = append(o.buf, b...)
}

func (o *O) WriteF64(v float64) {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, math.Float64bits(v))
	o.buf = append(o.buf, b...)
}

func (o *O) WriteCh(v byte) {
	o.buf = append(o.buf, v)
}

func (o *O) WriteCh8(v byte) {
	o.buf = append(o.buf, v)
}

func (o *O) WriteCh16(v uint16) {
	b := make([]byte, 2)
	binary.LittleEndian.PutUint16(b, v)
	o.buf = append(o.buf, b...)
}

func (o *O) WriteCh32(v uint32) {
	b := make([]byte, 4)
	binary.LittleEndian.PutUint32(b, v)
	o.buf = append(o.buf, b...)
}

func (o *O) WriteBl(v bool) {
	if v {
		o.buf = append(o.buf, 1)
	} else {
		o.buf = append(o.buf, 0)
	}
}

func (o *O) WriteSz(v int) {
	o.WriteU32(uint32(v))
}

func (o *O) WriteString(v string) {
	o.WriteSz(len(v))
	if len(v) > 0 {
		o.buf = append(o.buf, []byte(v)...)
	}
}

func (o *O) WriteU8String(v string) {
	o.WriteString(v)
}

func (o *O) WriteU16String(v string) {
	runes := []rune(v)
	o.WriteSz(len(runes))
	for _, r := range runes {
		b := make([]byte, 2)
		binary.LittleEndian.PutUint16(b, uint16(r))
		o.buf = append(o.buf, b...)
	}
}

func (o *O) WriteU32String(v []uint32) {
	o.WriteSz(len(v))
	for _, r := range v {
		b := make([]byte, 4)
		binary.LittleEndian.PutUint32(b, r)
		o.buf = append(o.buf, b...)
	}
}

func (o *O) WritePtrWithTypeID(value interface{}) {
	if value == nil {
		o.WriteU64(0)
		return
	}
	if obj, ok := value.(SpBase); ok {
		o.WriteU32(obj.TypeID())
		WriteObj(o, obj)
	} else {
		o.WriteU64(0)
	}
}

func (o *O) WritePtr(value interface{}, address uint64, writer func(interface{})) {
	if value == nil {
		o.WriteU64(0)
		return
	}
	o.WriteU64(address)
	if address != 0 {
		writer(value)
	}
}

func (o *O) WriteArray(arr []interface{}, writer func(interface{})) {
	o.WriteSz(len(arr))
	for _, v := range arr {
		writer(v)
	}
}

func (o *O) WriteSet(s map[interface{}]struct{}, writer func(interface{})) {
	o.WriteSz(len(s))
	for k := range s {
		writer(k)
	}
}

func (o *O) WriteMap(m map[interface{}]interface{}, keyWriter func(interface{}), valueWriter func(interface{})) {
	o.WriteSz(len(m))
	for k, v := range m {
		keyWriter(k)
		valueWriter(v)
	}
}

func (o *O) WriteSpArray(arr SpArray, writer func(interface{})) {
	for i := 0; i < arr.Size(); i++ {
		writer(arr.At(i))
	}
}

// =================== 类型分发（由 sp-gen 生成具体实现） ===================
// ReadObj / WriteObj 由 sp-gen 代码生成器产出，此处不定义桩函数
// 避免与生成器产出的实现冲突导致 Go 编译器报 redeclared 错误